File: python/string/trie.py
class Node:
    def __init__(self):
        self.common = 0
        self.nex = [-1]*26
        self.accept = False
        
class Trie:
    def __init__(self):
        self.char_size = 26
        self.base = 'a'
        self.nodes = [Node()]
    
    def insert(self, word):
        node_id = 0
        
        for wi in word:
            c = ord(wi)-ord(self.base)
            next_id = self.nodes[node_id].nex[c]
            
            if next_id==-1:
                next_id = len(self.nodes)
                self.nodes.append(Node())
                self.nodes[node_id].nex[c] = next_id
            
            self.nodes[node_id].common += 1
            node_id = next_id
        
        self.nodes[node_id].common += 1
        self.nodes[node_id].accept = True
    
    def search(self, word, prefix=False):
        node_id = 0
        
        for wi in word:
            c = ord(wi)-ord(self.base)
            next_id = self.nodes[node_id].nex[c]

            if next_id==-1:
                return False
            
            node_id = next_id
        
        return prefix or self.nodes[node_id].accept
    
    def count(self):
        return self.nodes[0].common

File: c++/string/trie.cpp
// https://leetcode.com/problems/implement-trie-prefix-tree/
class Trie {
public:
    struct Node {
        Node* nxt[26] = {};
        bool exist = false;
    };
    
    Node* root;
    
    Trie() {
        root = new Node;
    }
    
    void insert(string word) {
        Node* now = root;
        for (char c : word) {
            int x = c - 'a';
            if (now->nxt[x] == nullptr) now->nxt[x] = new Node;
            now = now->nxt[x];
        }
        now->exist = true;
    }
    
    bool search(string word) {
        Node* now = root;
        for (char c : word) {
            int x = c - 'a';
            if (now->nxt[x] == nullptr) return false;
            now = now->nxt[x];
        }
        return now->exist;
    }
    
    bool startsWith(string prefix) {
        Node* now = root;
        for (char c : prefix) {
            int x = c - 'a';
            if (now->nxt[x] == nullptr) return false;
            now = now->nxt[x];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
